Use nvm to keep independent node environments consistent

stevescavo opened the pull request:
This PR introduces Node Version Manager (nvm) as a way to better control our independent Node versions locally and otherwise. It does a preinstall check for the existence of nvm and then runs `nvm use` to pick up on the compatible version defined in `.nvmrc` before installing any dependencies.

stevescavo committed:
Require node 16, npm 8

stevescavo committed:
Source nvm (and node version) preinstall

stevescavo committed:
Check for nvm before running

stevescavo committed:
Create nvm.mjs script

stevescavo committed:
Change formatting

stevescavo committed:
Add more console documentation

claus commented on .nvmrc:
```diff
@@ -0,0 +1 @@
+v16.15.1
```
Is v16 LTS bound to a specific patch release?

stevescavo replied:
They typically are. Right now, [Node LTS](https://nodejs.org/) is actually 18.14.2.

claus replied:
It's just that 16.15.1 seems random. I'm running v16.13.0. `nvm ls` says LTS is v16.19.0.

claus replied:
Can we just put `v16`?

<img width="397" alt="Screenshot 2023-02-27 at 16 19 38" src="https://user-images.githubusercontent.com/2277/221661603-25383711-5dfb-4c4c-9c67-b56e79530e45.png">

stevescavo replied:
Yea, good call. I just changed `.nvmrc` to simply accept any minor or patched version of `v16`. I also removed the `engines: {}` definition in package.json because it's really not needed anymore and throws a loud warning right before `nvm` switches Node versions and resolves the issue :)

stevescavo committed:
Define Node 16 instead of specific version

claus reviewed (approved):
![lgtm1](https://user-images.githubusercontent.com/2277/221679723-921bf951-d977-4976-b1b2-366af41b87cc.gif)

stevescavo commented:
I'm now rethinking this due to the fact this is only a child process, so it can't update the node version in the parent shell. To update the version of Node in the parent process, we need to manually run `nvm use` in the parent shell session. I'm now working on only _**checking**_ for the required node version and then asking you update your node version using something like `nvm`.

stevescavo committed:
Check for required node version in package.json

stevescavo commented:
@claus in light of the rationale above, I've refactored this to simply check (warn, error or log) if the user has the correct Node version installed. It does nothing more because... we can't really.

stevescavo committed:
Remove comment

claus commented on scripts/check-node-version.mjs:
```diff
@@ -0,0 +1,33 @@
+const requiredNodeVersion = pkg.engines?.node;
+const currentVersion = process.versions.node;
+
+const requiredNodeVersionNumStr = requiredNodeVersion.replace(/[^\d.]/g, '');
+
+const requiredNodeVersionWholeNumber = Math.floor(
+    parseFloat(requiredNodeVersionNumStr)
+);
+const currentVersionWholeNumber = Math.floor(parseFloat(currentVersion));
```
I'll LGTM this but i'm gonna refactor this to use https://github.com/npm/node-semver

stevescavo replied:
Ah yea. Good call. I can just leave this open then?

claus reviewed (approved):
LGTM

claus committed:
Use node-semver to compare version numbers

Also loads `package.json` the traditional way, because `import asserts` is only supported in Node >=16.14. Michelle would have run into an exception.

I think the use of ESM should be okay, support is >=13.

claus commented:
@stevescavo Would you review my excellent code please, dear Sir?

stevescavo commented:
> @stevescavo Would you review my excellent code please, dear Sir?

@claus I have reviewed your excellent code and it's working excellently. Semver ftw.
